src/linear_algebra/matrix.lean

```diff
@@ -203,22 +204,33 @@ def diag : (matrix n n M) →ₗ[R] n → M := {
   (diag : (matrix n n R) →ₗ[R] n → R) 1 = λ i, 1 := by {
     dunfold diag, ext, simp [one_val_eq], }
 
-protected lemma smul_sum {α : Type u} {s : finset α} {f : α → M} {c : R} :
-  c • s.sum f = s.sum (λx, c • f x) := (s.sum_hom _).symm
+@[simp] lemma diag_transpose (A : matrix n n R) :
+  @diag n _ R R _ _ _ Aᵀ = @diag n _ R R _ _ _ A := rfl
 
 /--
 The trace of a square matrix.
 -/
 def trace : (matrix n n M) →ₗ[R] M := {
   to_fun := finset.univ.sum ∘ (diag : (matrix n n M) →ₗ[R] n → M),
   add    := by { intros, apply finset.sum_add_distrib, },
-  smul   := by { intros, simp [matrix.smul_sum], } }
+  smul   := by { intros, simp [finset.smul_sum], } }
 
 @[simp] lemma trace_one [decidable_eq n] :
   (trace : (matrix n n R) →ₗ[R] R) 1 = fintype.card n := by {
     have h : @trace n _ R R _ _ _ 1 = finset.univ.sum (@diag n _ R R _ _ _ 1) := rfl,
     rw [h, diag_one, finset.sum_const, add_monoid.smul_one], refl, }
 
+@[simp] lemma trace_transpose (A : matrix n n R) :
+  @trace n _ R R _ _ _ Aᵀ = @trace n _ R R _ _ _ A := rfl
```

Comment:
Since all the `@`s are needed here, it would probably be better to make the type arguments explicit in `trace`

Reply:
Agreed, I was beginning to think the same thing. I'll adjust.